test/test_foreach.py

```diff
@@ -77,23 +90,208 @@ def _get_test_data(self, device, dtype, N):
+                    inplace_ref(ref_inputs, **kwargs)
             else:
-                self.assertEqual(tensors1, expected)
+                inplace_ref(ref_inputs, **kwargs)
+                expected = ref_inputs[0]
+                if requires_cast:
+                    expected = [t.to(dtype) for t in expected]
+                if dtype in (torch.float16, torch.bfloat16) and TEST_WITH_ROCM:
+                    self.assertEqual(expected, inputs[0], atol=1.e-3, rtol=self.dtype_precisions[dtype][0])
+                else:
+                    self.assertEqual(expected, inputs[0])
+
+    def _test_binary_op_tensorlists(self, device, dtype, opinfo, N, is_fastpath, disable_fastpath):
+        n_expected_cudaLaunchKernels = N if disable_fastpath else 1
+        op, ref, inplace_op, inplace_ref = self._get_funcs(opinfo, n_expected_cudaLaunchKernels)
+        inputs = [
+            opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath),
+            opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath),
+        ]
+        self._regular_binary_test(dtype, op, ref, inputs, is_fastpath)
+        self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath)
+        if opinfo.supports_alpha_param:
+            alpha = 3 if dtype in torch.testing.get_all_int_dtypes() else 3.14
+            self._regular_binary_test(dtype, op, ref, inputs, is_fastpath, alpha=alpha)
+            self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath, alpha=alpha)
+
+    @skipMeta
+    @ops(foreach_binary_op_db)
+    def test_binary_op_tensorlists_fastpath(self, device, dtype, op):
+        for N in N_values:
+            disable_fastpath = op.ref == torch.div and dtype in torch.testing.get_all_int_dtypes() + [torch.bool]
+            if op.ref == torch.add and dtype == torch.bool:
+                disable_fastpath = True
+            self._test_binary_op_tensorlists(device, dtype, op, N, True, disable_fastpath)
+
+    @dtypes(*torch.testing.get_all_dtypes())
+    @ops(foreach_binary_op_db)
+    def test_binary_op_tensorlists_slowpath(self, device, dtype, op):
+        for N in N_values:
+            self._test_binary_op_tensorlists(device, dtype, op, N, False, False)
+
+    def _test_binary_op_scalar(self, device, dtype, opinfo, N, scalar, is_fastpath, disable_fastpath):
+        n_expected_cudaLaunchKernels = N if disable_fastpath else 1
+        op, ref, inplace_op, inplace_ref = self._get_funcs(opinfo, n_expected_cudaLaunchKernels)
+        inputs = [opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath), scalar]
+        self._regular_binary_test(dtype, op, ref, inputs, is_fastpath)
+        self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath)
+
+    @skipCUDAIfRocm
+    @skipMeta
+    @ops(foreach_binary_op_db)
+    def test_binary_op_scalar_fastpath(self, device, dtype, op):
+        for N in N_values:
```

Comment:
You can have `for N, scalar in product(N_values, scalars)`:
(you can create `scalars` tuple in advance)